### PR TITLE
[SPARK-39504][CORE][SQL] No-null object invokes equals method to avoid NPE

### PR DESCRIPTION
**Pull request by dcoliversun**
### What changes were proposed in this pull request?

This PR aims to make `equals` method invoked by No-null object to avoid NPE.

### Why are the changes needed?

Since NullPointerException can possibly be thrown while calling the equals method of Object, equals should be invoked by a constant or an object that is definitely not null.

> Positive example: "test".equals(object);
> Counter example: object.equals("test");

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

No need to add tests because this PR just changes object invoked.

### CODE REVIEW

File: common/network-common/src/main/java/org/apache/spark/network/TransportContext.java [hunk 1]
```diff
@@ -124,7 +124,7 @@ public TransportContext(
     this.closeIdleConnections = closeIdleConnections;
 
     if (conf.getModuleName() != null &&
-        conf.getModuleName().equalsIgnoreCase("shuffle") &&
+        "shuffle".equalsIgnoreCase(conf.getModuleName()) &&
```
Comment (LuciferYang) [inline common/network-common/src/main/java/org/apache/spark/network/TransportContext.java:127]: hmm... if `conf.getModuleName() != null` is true,  I think NPE won't happen here

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java [hunk 2]
```diff
@@ -306,7 +306,7 @@ public long maxChunksBeingTransferred() {
    * ceiling of the nearest integer.
    */
   public int chunkFetchHandlerThreads() {
-    if (!this.getModuleName().equalsIgnoreCase("shuffle")) {
+    if (!"shuffle".equalsIgnoreCase(this.getModuleName())) {
```
Comment (LuciferYang) [inline common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java:309]: https://github.com/apache/spark/blob/75b396505a266e839b9e2dc1d3ec36eebb6deb43/common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java#L51-L55

Maybe we should change line 55 to `this.module = Objects.requireNonNull(module);` to ensure `this.module` is not null,

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java [hunk 3]
```diff
@@ -394,7 +394,7 @@ private class ManagedBufferIterator implements Iterator<ManagedBuffer> {
         size = reduceIdAndChunkIds.length;
         blockDataForIndexFn = index -> mergeManager.getMergedBlockData(msg.appId, shuffleId,
           shuffleMergeId, reduceIdAndChunkIds[index], reduceIdAndChunkIds[index + 1]);
-      } else if (blockId0Parts.length == 3 && blockId0Parts[0].equals("rdd")) {
+      } else if (blockId0Parts.length == 3 && "rdd".equals(blockId0Parts[0])) {
```
Comment (LuciferYang) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:397]: In what scenario `blockId0Parts[0]` is null?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java [hunk 4]
```diff
@@ -319,7 +319,7 @@ public static void throwException(Throwable t) {
   static {
     boolean _unaligned;
     String arch = System.getProperty("os.arch", "");
-    if (arch.equals("ppc64le") || arch.equals("ppc64") || arch.equals("s390x")) {
+    if ("ppc64le".equals(arch) || "ppc64".equals(arch) || "s390x".equals(arch)) {
```
Comment (LuciferYang) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:322]: In what scenario`System.getProperty("os.arch", ""` is null?

File: sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/thrift/ThriftCLIService.java [hunk 5]
```diff
@@ -282,8 +282,8 @@ private String getIpAddress() {
     String clientIpAddress;
     // Http transport mode.
     // We set the thread local ip address, in ThriftHttpServlet.
-    if (cliService.getHiveConf().getVar(
-        ConfVars.HIVE_SERVER2_TRANSPORT_MODE).equalsIgnoreCase("http")) {
+    if ("http".equalsIgnoreCase(cliService.getHiveConf().getVar(
```
Comment (LuciferYang) [inline sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/thrift/ThriftCLIService.java:285]: This file copy from `hive`, It's best not to touch it to avoid conflict

File: sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/thrift/ThriftHttpCLIService.java [hunk 6]
```diff
@@ -194,7 +194,7 @@ public void run() {
    * @return
    */
   private String getHttpPath(String httpPath) {
-    if(httpPath == null || httpPath.equals("")) {
+    if(httpPath == null || httpPath.isEmpty()) {
```
Comment (LuciferYang) [inline sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/thrift/ThriftHttpCLIService.java:197]: ditto

File: sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/thrift/ThriftHttpServlet.java [hunk 7]
```diff
@@ -553,7 +553,7 @@ private static String getDoAsQueryParam(String queryString) {
     Map<String, String[]> params = javax.servlet.http.HttpUtils.parseQueryString( queryString );
     Set<String> keySet = params.keySet();
     for (String key: keySet) {
-      if (key.equalsIgnoreCase("doAs")) {
+      if ("doAs".equalsIgnoreCase(key)) {
```
Comment (LuciferYang) [inline sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/thrift/ThriftHttpServlet.java:556]: ditto

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java [hunk 8]
```diff
@@ -128,7 +128,7 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
     TaskContext taskContext = TaskContext$.MODULE$.get();
     if (taskContext != null) {
       Option<AccumulatorV2<?, ?>> accu = taskContext.taskMetrics().externalAccums().lastOption();
-      if (accu.isDefined() && accu.get().getClass().getSimpleName().equals("NumRowGroupsAcc")) {
+      if (accu.isDefined() && "NumRowGroupsAcc".equals(accu.get().getClass().getSimpleName())) {
```
[the remark this answers is not in the document]
Comment (dcoliversun) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java]: ok, back off

File: common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java (hunk 4 above)
Comment (dcoliversun) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java]: no scenario. back off

File: common/network-common/src/main/java/org/apache/spark/network/TransportContext.java (hunk 1 above)
Comment (dcoliversun) [reply common/network-common/src/main/java/org/apache/spark/network/TransportContext.java]: remove not null logic

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java (hunk 3 above)
Comment (dcoliversun) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:397]: Just defensive programming. It's possible that `blockId0Parts[0]` is null and no possible that `"rdd"` is null

File: sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/thrift/ThriftCLIService.java (hunk 5 above)
Comment (dcoliversun) [reply sql/hive-thriftserver/src/main/java/org/apache/hive/service/cli/thrift/ThriftCLIService.java:285]: OK. If these are copied from Apache Hive, it's better to submit PR to Hive and cherry pick

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java (hunk 2 above)
Comment (dcoliversun) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: OK, it's better

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtils.java [hunk 9]
```diff
@@ -62,13 +62,13 @@ private static byte[] aesInternal(
...
-          (padding.equalsIgnoreCase("PKCS") || padding.equalsIgnoreCase("DEFAULT"))) {
+      if ("ECB".equalsIgnoreCase(mode) &&
+          ("PKCS".equalsIgnoreCase(padding) || "DEFAULT".equalsIgnoreCase(padding))) {
         Cipher cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
         cipher.init(opmode, secretKey);
         return cipher.doFinal(input, 0, input.length);
-      } else if (mode.equalsIgnoreCase("GCM") &&
-          (padding.equalsIgnoreCase("NONE") || padding.equalsIgnoreCase("DEFAULT"))) {
+      } else if ("GCM".equalsIgnoreCase(mode) &&
+          ("NONE".equalsIgnoreCase(padding) || "DEFAULT".equalsIgnoreCase(padding))) {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtils.java:71]: Could you revert these changes of this file, @dcoliversun ?
This is logically impossible because `aesInternal` is `private static` and used in the constructor with `mode.toString()`.

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java [hunk 10]
```diff
@@ -401,8 +401,7 @@ boolean isClientMode(Map<String, String> userProps) {
    * Return whether the given main class represents a thrift server.
    */
   private boolean isThriftServer(String mainClass) {
-    return (mainClass != null &&
-      mainClass.equals("org.apache.spark.sql.hive.thriftserver.HiveThriftServer2"));
+    return ("org.apache.spark.sql.hive.thriftserver.HiveThriftServer2".equals(mainClass));
```
Comment (dongjoon-hyun) [inline launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java:404]: Is this needed? It looks like we already have some defensive code for this `private boolean isThriftServer`.

https://github.com/apache/spark/blob/cb0ef06b641845be5774b18c9c24d67c6ab4561a/launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java#L235

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java (hunk 3 above)
Comment (dongjoon-hyun) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java:397]: Unfortunately, it doesn't looks like this code has any improvement like you claimed, @dcoliversun .
If `blockID0Parts[0] == null`, it fails already at line 383 before reaches this line 397.

File: common/network-common/src/main/java/org/apache/spark/network/TransportContext.java [hunk 11]
```diff
@@ -123,8 +123,7 @@ public TransportContext(
     this.rpcHandler = rpcHandler;
     this.closeIdleConnections = closeIdleConnections;
 
-    if (conf.getModuleName() != null &&
-        conf.getModuleName().equalsIgnoreCase("shuffle") &&
+    if ("shuffle".equalsIgnoreCase(conf.getModuleName()) &&
```
Comment (dongjoon-hyun) [inline common/network-common/src/main/java/org/apache/spark/network/TransportContext.java:126]: Previously, there is no NPE in this case. So, this is just a code-implication instead of avoiding any NPE.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I left a few comments on the invalid or mismatched codes, @dcoliversun . Please revert them.

The only part left is `TransportConf`, but we need to dig more about the real contribution on that part.

**Comment by dongjoon-hyun** [comment]
BTW, thank you for your review, @LuciferYang !

**Comment by dcoliversun** [comment]
@dongjoon-hyun Thanks for your advise. As you said, This PR is no longer about no-null object invokes equals to avoid NPE. I will close this PR and submit a new PR.

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java (hunk 10 above)
Comment (dcoliversun) [reply launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java:404]: It's needed. If remove `mainClass != null`, pyspark test will fail.
```plain
========================================================================
Running PySpark tests
========================================================================
Running PySpark tests. Output is in /__w/spark/spark/python/unit-tests.log
Will test against the following Python executables: ['python3.9', 'pypy3']
Will test the following Python modules: ['pyspark-sql', 'pyspark-resource', 'pyspark-mllib']
python3.9 python_implementation is CPython
python3.9 version is: Python 3.9.5
pypy3 python_implementation is PyPy
pypy3 version is: Python 3.7.12 (7.3.7+dfsg-1~ppa1~ubuntu20.04, Oct 26 2021, 02:00:43)
[PyPy 7.3.7 with GCC 9.3.0]
Starting test(pypy3): pyspark.sql.tests.test_arrow (temp output: /tmp/pypy3__pyspark.sql.tests.test_arrow__2_kthvri.log)
Finished test(pypy3): pyspark.sql.tests.test_arrow (1s) ... 121 tests were skipped
Starting test(pypy3): pyspark.sql.tests.test_arrow_map (temp output: /tmp/pypy3__pyspark.sql.tests.test_arrow_map__ot8ujcj9.log)
Finished test(pypy3): pyspark.sql.tests.test_arrow_map (0s) ... 7 tests were skipped
Starting test(pypy3): pyspark.sql.tests.test_catalog (temp output: /tmp/pypy3__pyspark.sql.tests.test_catalog__67s8rrv3.log)
Exception in thread "main" java.lang.NullPointerException
	at org.apache.spark.launcher.SparkSubmitCommandBuilder.isThriftServer(SparkSubmitCommandBuilder.java:404)
	at org.apache.spark.launcher.SparkSubmitCommandBuilder.buildSparkSubmitCommand(SparkSubmitCommandBuilder.java:265)
	at org.apache.spark.launcher.SparkSubmitCommandBuilder.buildCommand(SparkSubmitCommandBuilder.java:164)
	at org.apache.spark.launcher.Main.buildCommand(Main.java:113)
	at org.apache.spark.launcher.Main.main(Main.java:63)
```
